app/controllers/PayPeriodQuestionController.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import controllers.actions._
+import forms.PayPeriodQuestionFormProvider
+import javax.inject.Inject
+import models.Mode
+import navigation.Navigator
+import pages.{PayDatePage, PayPeriodQuestionPage}
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import repositories.SessionRepository
+import uk.gov.hmrc.play.bootstrap.controller.FrontendBaseController
+import views.html.PayPeriodQuestionView
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class PayPeriodQuestionController @Inject()(
+  override val messagesApi: MessagesApi,
+  sessionRepository: SessionRepository,
+  navigator: Navigator,
+  identify: IdentifierAction,
+  getData: DataRetrievalAction,
+  requireData: DataRequiredAction,
+  formProvider: PayPeriodQuestionFormProvider,
+  val controllerComponents: MessagesControllerComponents,
+  view: PayPeriodQuestionView
+)(implicit ec: ExecutionContext)
+    extends FrontendBaseController with I18nSupport {
+
+  val form = formProvider()
+
+  def onPageLoad(): Action[AnyContent] = (identify andThen getData andThen requireData) { implicit request =>
+    val preparedForm = request.userAnswers.get(PayPeriodQuestionPage) match {
+      case None        => form
+      case Some(value) => form.fill(value)
+    }
+
+    val payDates = request.userAnswers.getList(PayDatePage)
+    Ok(view(preparedForm, payDates.sliding(2)))
+  }
+
+  def onSubmit(): Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
+    form
+      .bindFromRequest()
+      .fold(
+        formWithErrors => {
+          val payDates = request.userAnswers.getList(PayDatePage)
+          Future.successful(BadRequest(view(formWithErrors, payDates.sliding(2))))
```

Comment:
If you extend PeriodHelper you can use a function called generatePeriods which will give you a list of pay periods

this will keep the +1 days logic out of the view

Reply:
looks like I need to extend `FurloughPeriodExtractor` too in addition to `PeriodHelper`

Reply:
the furlough period is only needed to detect full or partial periods, you could breakdown the generatePeriods function, split the tail recursion bit into a separate function which only requires Seq[LocalDate] and then call that from generatePeriods

Reply:
looking...